src/System.Private.CoreLib/shared/System/Collections/Concurrent/ConcurrentQueue.cs

```diff
@@ -304,29 +307,63 @@ public int Count
+                        int tailTail = Volatile.Read(ref tail._headAndTail.Tail);
                         if (head == _head &&
                             tail == _tail &&
                             headHead == Volatile.Read(ref head._headAndTail.Head) &&
                             headTail == Volatile.Read(ref head._headAndTail.Tail) &&
                             tailHead == Volatile.Read(ref tail._headAndTail.Head) &&
                             tailTail == Volatile.Read(ref tail._headAndTail.Tail))
                         {
-                            // We got stable values, so we can just compute the sizes based on those
-                            // values and return the sum of the counts of the segments.
                             return GetCount(head, headHead, headTail) + GetCount(tail, tailHead, tailTail);
                         }
                     }
                     else
                     {
-                        // There were more than two segments.  Take the slower path, where we freeze the
-                        // queue and then count the now stable segments.
-                        SnapForObservation(out head, out headHead, out tail, out tailTail);
-                        return unchecked((int)GetCount(head, headHead, tail, tailTail));
+                        // There were more than two segments in the queue.  Fall back to taking the cross-segment lock,
+                        // which will ensure that the head and tail segments we read are stable (since the lock is needed to change them);
+                        // for the two-segment case above, we can simply rely on subsequent comparisons, but for the two+ case, we need
+                        // to be able to trust the internal segments between the head and tail.
+                        lock (_crossSegmentLock)
+                        {
+                            // Now that we hold the lock, re-read the previously captured head and tail segments and head positions.
+                            // If either has changed, start over.
+                            if (head == _head && tail == _tail)
+                            {
+                                // Get the positions from the tail, and as above, if the captured values match the previous reads,
+                                // we can use the values to compute the count of the head and tail segments.
+                                int tailHead = Volatile.Read(ref tail._headAndTail.Head);
+                                int tailTail = Volatile.Read(ref tail._headAndTail.Tail);
+                                if (headHead == Volatile.Read(ref head._headAndTail.Head) &&
+                                    headTail == Volatile.Read(ref head._headAndTail.Tail) &&
+                                    tailHead == Volatile.Read(ref tail._headAndTail.Head) &&
+                                    tailTail == Volatile.Read(ref tail._headAndTail.Tail))
```

Comment:
Is it possible for the double-checking on the head and tail segments' head and tail here and above to cause this API to spin for a very long time if enqueue/dequeue are happening very actively?

Would it be better to:
- Double-check for 1 and 2 segment cases as above once
- But on failure of initial condition or the double-check, fall back to lock and inside the lock, don't double-check
Typically the lock should not be contended (once the queue stabilizes in size and the tail does not change frequently)

Reply:
> Is it possible for the double-checking on the head and tail segments' head and tail here and above to cause this API to spin for a very long time if enqueue/dequeue are happening very actively?

Yes, it's possible. (That's true today as well.)

> inside the lock, don't double-check

Can you explain this?

Reply:
> Yes, it's possible. (That's true today as well.)

For the 1 and 2 segment cases it's true today, but now it's also true when there are multiple segments (though that's a less interesting case). I wonder if even that is necessary. Count only provides a snapshot and it can't be relied upon without appropriate synchronization by callers, why double-check at all?

> inside the lock, don't double-check

I mean inside the lock, can we just read the values and compute the counts without double-checking new values with previous values?

Generally wondering if the spinning can be removed altogether.

Reply:
> Aren't those true despite the double-checking? I figured the double-checking would only make it less likely but not impossible.

For it to not be true there would have to be more ~4 billion items enqueued and dequeued between the checks, in order for the values to wrap around.

Reply:
One possible thing we could do in the locked 2+ case is have a loop within the lock.  While holding the lock, the only operations that can be performed are dequeueing from the head segment and enqueueing to the tail segment, which means the max number of operations that can occur is the number of currently queued items in the head segment and the number of empty spaces in the tail segment.  That's a finite number, worst case 2M enqueues and dequeues due to MaxSegmentLength being 1M.  But, it's still a very big number.

Reply:
Ah ok I see how that works, makes sense